src/operator/setup.rs

```diff
@@ -0,0 +1,185 @@
+    pub fn apply_command<C>(&self, command: &C) -> Result<C>
+    where
+        C: Clone + Debug + DeserializeOwned + Resource<DynamicType = ()> + Serialize,
+    {
+        let cmd: C = self.client.apply(&serde_yaml::to_string(command)?);
+
+        // we wait here to give the operator time to react to the command
+        thread::sleep(Duration::from_secs(2));
+        Ok(cmd)
+    }
+
+    /// Deletes the custom resource, and waits for pods to be terminated
+    /// This should be executed after every single test to provide a "clean"
+    /// environment for the following tests.
+    pub fn delete(&mut self) -> Result<()> {
+        if let Some(cluster) = self.cluster.take() {
+            self.client.delete(cluster);
+            self.wait_for_pods_terminated()?;
+            self.cluster = None;
+        }
+
+        Ok(())
+    }
+
+    /// Delete a command resource. This is for clean up purposes.
+    pub fn delete_command<C>(&mut self, command: C) -> Result<()>
+    where
+        C: Clone + Debug + DeserializeOwned + Resource<DynamicType = ()> + Serialize,
+    {
+        self.client.delete(command);
+        Ok(())
+    }
+
+    /// Returns all available pods in the cluster via the name label.
+    pub fn get_current_pods(&self) -> Vec<Pod> {
+        let current_pods: ObjectList<Pod> = self.client.list_labeled(&format!(
+            "app.kubernetes.io/name={}",
+            &self.options.pod_name_label
+        ));
+        current_pods.items
+    }
+
+    /// Returns all available nodes in the cluster. Can be used to determine the expected pods
+    /// for tests (depending on the custom resource)
+    pub fn get_available_nodes(&self) -> Vec<Node> {
+        let available_nodes: ObjectList<Node> = self.client.list_labeled("");
+        available_nodes.items
+    }
+
+    /// A "busy" wait for all pods to be terminated and cleaned up.
+    pub fn wait_for_pods_terminated(&self) -> Result<()> {
+        let now = Instant::now();
+
+        while now.elapsed().as_secs() < self.timeouts.pods_terminated.as_secs() {
+            let pods = self.get_current_pods();
+
+            if pods.is_empty() {
+                return Ok(());
+            }
+
+            println!("Waiting for {} Pod(s) to terminate", pods.len());
+            thread::sleep(Duration::from_secs(1));
+        }
+
+        Err(anyhow!(
+            "Pods did not terminate within the specified timeout of {} second(s)",
+            self.timeouts.pods_terminated.as_secs()
+        ))
+    }
+}
+
+impl<T> TestCluster<T>
+where
+    T: Clone + Conditions + Debug + DeserializeOwned + Resource<DynamicType = ()> + Serialize,
+{
+    /// Creates or updates a custom resource and waits for the cluster to be up and running
+    /// within the provided timeout. Depending on the cluster definition we hand in the number
+    /// of created pods we expect manually.
+    pub fn create_or_update(&mut self, cluster: &T, expected_pod_count: usize) -> Result<()> {
```

Comment:
Ok. So a clean environment is assumend. This is okay for me.

Reply:
Yes its assumed and should be like that after adding the drop trait.